lib/hpke/src/certificate.cpp

```diff
@@ -25,14 +39,104 @@ struct Certificate::ParsedCertificate
+    std::string skid_str(reinterpret_cast<const char*>(octet_skid));
+    return bytes(skid_str.begin(), skid_str.end());
+  }
+
+  // Parse Authority Key Identifier
+  static bytes parse_akid(X509* cert)
+  {
+    const auto* asn_akid = X509_get0_authority_key_id(cert);
+    if (asn_akid == nullptr) {
+      return bytes{};
+    }
+    const auto* octet_akid = ASN1_STRING_get0_data(asn_akid);
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    std::string akid_str(reinterpret_cast<const char*>(octet_akid));
+    return bytes(akid_str.begin(), akid_str.end());
+  }
+
+  static std::vector<GeneralName> parse_san(X509* cert)
+  {
+    std::vector<GeneralName> san_info;
+    int san_names_nb = -1;
+
+    const auto san_names =
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      make_typed_unique(reinterpret_cast<STACK_OF(GENERAL_NAME)*>(
+        X509_get_ext_d2i(cert, NID_subject_alt_name, nullptr, nullptr)));
+    san_names_nb = sk_GENERAL_NAME_num(san_names.get());
+
+    // Check each name within the extension
+    for (int i = 0; i < san_names_nb; i++) {
+      auto* current_name = sk_GENERAL_NAME_value(san_names.get(), i);
+
+      if (current_name->type == GEN_DNS) {
+        const char* dns_name =
+          // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast)
+          reinterpret_cast<const char*>((ASN1_STRING_get0_data(
+            current_name->d
+              .dNSName))); // NOLINT(cppcoreguidelines-pro-type-union-access)
+
+        // Make sure there isn't an embedded NUL character in the DNS name
+        if (ASN1_STRING_length(
+              current_name->d
+                .dNSName) != // NOLINT(cppcoreguidelines-pro-type-union-access)
+            static_cast<int>(strlen(dns_name))) {
+          throw std::runtime_error("Malformed certificate");
+        }
+        san_info.emplace_back(DNSName{ dns_name });
+
+      } else if (current_name->type == GEN_EMAIL) {
+        const char* email =
+          // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast)
+          reinterpret_cast<const char*>(ASN1_STRING_get0_data(
+            current_name->d
+              .dNSName)); // NOLINT(cppcoreguidelines-pro-type-union-access)
+
+        // Make sure there isn't an embedded NUL character in the DNS name
+        if (ASN1_STRING_length(
+              current_name->d
+                .dNSName) != // NOLINT(cppcoreguidelines-pro-type-union-access)
+            static_cast<int>(strlen(email))) {
+          throw std::runtime_error("Malformed certificate");
+        }
+        san_info.emplace_back(RFC822Name{ email });
+      }
+    }
+
+    return san_info;
+  }
+
   explicit ParsedCertificate(X509* x509_in)
     : x509(x509_in, typed_delete<X509>)
     , sig_id(signature_algorithm(x509.get()))
+    , issuer(X509_NAME_oneline(X509_get_issuer_name(x509.get()), nullptr, 0))
+    , subject(X509_NAME_oneline(X509_get_subject_name(x509.get()), nullptr, 0))
```

Comment:
These are still wrong.  My guess is that you want one of two patterns here:

1. `X509_NAME_get0_der(X509_get_subject_name(x509))`
2. `X509_subject_name_hash(x509)`

I think you want the latter, since you're only using the subject and issuer for comparison.  That means that `issue` and `subject` can just be long integers (let's use uint64_t to be explicit).